### PR TITLE
`Integrated code lifecycle`: Prevent pushes to non-default branches

### PR DESCRIPTION
**Pull request by laurenzfb**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I adapted an integration test to account for the new changes.
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Artemis currently does not support the use of other branches. Users should not be able to create new branches in Local VC.

### Description

This PR adds a condition in the Local VC Pre Push Hook to prevent pushes to branches other than `main`. Previously, those pushes were allowed and saved but otherwise ignored.

### Steps for Testing

**Test on TS3, TS6 or Local CI Staging**

1. Clone a programming exercise repository locally
2. Create a new branch and push it, e.g. `git checkout -b test`, then `git push --set-upstream origin test`
3. You should receive the following error message: "You cannot push to a branch other than the default branch."

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ArtemisGitServlet.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6099/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.config.localvcci/ArtemisGitServlet.html) | 100% | ✅ |
| [LocalVCPrePushHook.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6099/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localvc/LocalVCPrePushHook.html) | 81% | ✅ |
| [LocalVCService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6099/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localvc/LocalVCService.html) | 73% | ✅ |
| [LocalVCServletService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6099/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localvc/LocalVCServletService.html) | 91% | ✅ |

## Summary by CodeRabbit

- **New Features**
	- Enhanced version control integration to enforce push operations to the default branch only, improving workflow consistency.
- **Bug Fixes**
	- Fixed the issue where users could push to non-default branches.
- **Tests**
	- Updated integration tests to cover new version control behavior and branch restrictions.

### CODE REVIEW

**Comment by Strohgelaender** [comment]
Not sure if we should really permit pushes to other branches. While it's true that the rated build will only be triggered by the main branch, other branches might be helpful when working on team exercises or when multiple editors are creating test cases.

**Comment by Hialus** [comment]
I agree with Lucas that in general having other branches is useful. However, Artemis currently does not provide any UI for different branches (and PRs for that matter). This could then result in pushes to other branches that are hard to trace, unless you have the repo locally and know what to look for (which students in general do not).
And apart from team exercises having branches was also useful for already running homework exercises to propose changes (e.g. bugfixes) as an editor and getting them approved by an instructor.

As it stands now I'd support this solution to prevent confusion. But ideally Artemis would in the future provide proper UIs for branch management (and ideally even PRs). This could be useful for both team exercises, but also exercise creators.

**Comment by laurenzfb** [comment]
@Strohgelaender I discussed with Stephan and we decided to move forward with preventing pushes to non-default branches for now. We will enable it again in the future, once Artemis properly supports this feature.

I also refactored the prevention by actually checking the default branch of the repository and not using hardcoded main

**Review by pzdr7 — APPROVED** [review]
Tested on TS3. Works as expected - I get the correct error message when trying to push to the remote (e.g. `git push origin main:foo`)

**Review by milljoniaer — APPROVED** [review]
Tested on ts3, works as expected 🚀

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java [hunk 1]
```diff
@@ -501,4 +504,16 @@ private Commit extractCommitInfo(String commitHash, Repository repository) throw
...
         return new Commit(commitHash, author.getName(), revCommit.getFullMessage(), author.getEmailAddress(), branch);
     }
+
+    /**
+     * Determine the default branch of the given repository.
+     *
+     * @param repository the repository for which the default branch should be determined.
+     * @return the name of the default branch.
+     */
+    public String getMainBranchOfRepository(Repository repository) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java:514]: consistent naming.
```suggestion
    public String getDefaultBranchOfRepository(Repository repository) {
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCPrePushHook.java [hunk 2]
```diff
@@ -44,6 +52,21 @@ public void onPreReceive(ReceivePack receivePack, Collection<ReceiveCommand> com
...
+        try {
+            mainBranchName = localVCServletService.getMainBranchOfRepository(repository);
+        }
+        catch (LocalVCInternalException e) {
+            command.setResult(ReceiveCommand.Result.REJECTED_OTHER_REASON, "An error occurred while checking the branch.");
+            return;
+        }
+
+        // Reject pushes to anything other than the default branch.
+        if (!command.getRefName().equals("refs/heads/" + mainBranchName)) {
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCPrePushHook.java:65]: Further aligning the naming mentioned by Stroh.
```suggestion
        String defaultBranchName;
        try {
            defaultBranchName = localVCServletService.getDefaultBranchOfRepository(repository);
        }
        catch (LocalVCInternalException e) {
            command.setResult(ReceiveCommand.Result.REJECTED_OTHER_REASON, "An error occurred while checking the branch.");
            return;
        }

        // Reject pushes to anything other than the default branch.
        if (!command.getRefName().equals("refs/heads/" + defaultBranchName)) {
```

**Review by rstief — COMMENTED** [review]
looks good overall, just adding to the re-naming.

**Review by pzdr7 — APPROVED** [review]
Reapprove

**Review by milljoniaer — APPROVED** [review]

**Review by reschandreas — APPROVED** [review]
code lgtm

**Review by rstief — APPROVED** [review]

**Review by bassner — APPROVED** [review]
LGTM

**Pull request merged**